controller/program/src/instructions/liquidity/add.rs

```diff
@@ -0,0 +1,368 @@
+    abr: &mut Abr,
+    accounts: &[AccountHandle],
+    ix_args: AddLiquidityIxArgs,
+    cpi: &mut Cpi,
+) -> Result<(), ProgramError> {
+    let AddLiquidityIxAccounts {
+        ix_prefix,
+        lst_calc_prog,
+        lst_calc,
+        pricing_prog,
+        pricing,
+    } = add_liquidity_accs_checked(abr, accounts, ix_args)?;
+
+    let lp_lst_supply = RawTokenAccount::of_acc_data(abr.get(*ix_prefix.pool_reserves()).data())
+        .and_then(TokenAccount::try_from_raw)
+        .map(|a| a.amount())
+        .ok_or(INVALID_ACCOUNT_DATA)?;
+
+    lst_sync_sol_val_unchecked(
+        abr,
+        cpi,
+        SyncSolValueIxAccs {
+            ix_prefix: NewSyncSolValueIxPreAccsBuilder::start()
+                .with_lst_mint(*ix_prefix.lst_mint())
+                .with_pool_state(*ix_prefix.pool_state())
+                .with_lst_state_list(*ix_prefix.lst_state_list())
+                .with_pool_reserves(*ix_prefix.pool_reserves())
+                .build(),
+            calc_prog: lst_calc_prog,
+            calc: lst_calc,
+        },
+        ix_args.lst_index as usize,
+    )?;
+
+    // Extract the data you need from pool before CPI calls
+    let start_total_sol_value = unsafe {
+        PoolState::of_acc_data(abr.get(*ix_prefix.pool_state()).data())
+            .ok_or(Inf1CtlCustomProgErr(Inf1CtlErr::InvalidPoolStateData))?
+            .total_sol_value
+    };
+
+    // Step 4: Calculate sol_value_to_add = LstToSol(amount).min
+
+    let lst_amount_sol_value = LstToSolRetVal(cpi_lst_to_sol(
+        cpi,
+        abr,
+        lst_calc_prog,
+        ix_args.amount,
+        SvcIxAccountHandles::new(
+            NewSvcIxPreAccsBuilder::start()
+                .with_lst_mint(*ix_prefix.lst_mint())
+                .build(),
+            lst_calc,
+        ),
+    )?);
+
+    // Step 5: Calculate sol_value_to_add_after_fees = PriceLpTokensToMint(lp_tokens_sol_value)
+    let lst_amount_sol_value_after_fees = PricingRetVal(cpi_price_exact_in(
+        cpi,
+        abr,
+        pricing_prog,
+        PriceExactInIxArgs {
+            sol_value: *lst_amount_sol_value.0.end(),
+            amt: ix_args.amount,
+        },
+        PriceInIxAccountHandles::new(
+            NewPPIxPreAccsBuilder::start()
+                .with_input_mint(*ix_prefix.lst_mint())
+                .with_output_mint(*ix_prefix.lp_token_mint())
+                .build(),
+            pricing,
+        ),
+    )?);
+
+    let pool = unsafe { PoolState::of_acc_data(abr.get(*ix_prefix.pool_state()).data()) }
+        .ok_or(Inf1CtlCustomProgErr(Inf1CtlErr::InvalidPoolStateData))?;
+
+    // Will dilute existing LPs if unchecked
+    if lst_amount_sol_value_after_fees.0 > *lst_amount_sol_value.0.end() {
+        return Err(Inf1CtlCustomProgErr(Inf1CtlErr::PoolWouldLoseSolValue).into());
+    }
+
+    let add_liquidity_quote = match quote_add_liq(AddLiqQuoteArgs {
+        amt: ix_args.amount,
+        lp_token_supply: lp_lst_supply,
```

Comment:
i dont even know how your tests are passing right now. Did you just coincidentally have these 2 values as the same in the fixtures?